File: rust_exercises/leetcode/Problem733/solution.rs
impl Solution {
    pub fn flood_fill(image: Vec<Vec<i32>>, sr: i32, sc: i32, new_color: i32) -> Vec<Vec<i32>> {
        let mut image = image;
        let old_color = image[sr as usize][sc as usize];
        if old_color == new_color {
            return image;
        }
        dfs(&mut image, sr, sc, old_color, new_color);
        image
    }
}

fn dfs(image: &mut Vec<Vec<i32>>, r: i32, c: i32, old_color: i32, new_color: i32) {
    image[r as usize][c as usize] = new_color;
    
    let moves = [(1, 0), (-1, 0), (0, 1), (0, -1)];
    for (mr, mc) in moves.iter() {
        let nextr = r + mr;
        let nextc = c + mc;
        
        if valid(nextr, nextc, image.len(), image[0].len()) && image[nextr as usize][nextc as usize] == old_color {
            dfs(image, nextr, nextc, old_color, new_color);
        }
    }
}

fn valid(i: i32, j: i32, m: usize, n: usize) -> bool {
    i >= 0 && j >= 0 && i < m as i32 && j < n as i32
}

File: rust_exercises/leetcode/Problem733/Solution.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int oldColor = image[sr][sc];
        if (oldColor == newColor) {
            return image;
        }
        dfs(image, sr, sc, oldColor, newColor);
        return image;
    }
    
    private void dfs(int[][] image, int r, int c, int oldColor, int newColor) {
        image[r][c] = newColor;
        int[][] moves = new int[][]{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        for (int k = 0; k < 4; k += 1) {
            int nextr = r + moves[k][0];
            int nextc = c + moves[k][1];
            if (valid(nextr, nextc, image.length, image[0].length) && image[nextr][nextc] == oldColor) {
                dfs(image, nextr, nextc, oldColor, newColor);
            }
        }
    }
    
    private boolean valid(int i, int j, int m, int n) {
        return i >= 0 && j >= 0 && i < m && j < n;
    }
}
